aws-lc-rs/src/unstable/kdf/sskdf.rs

```diff
@@ -0,0 +1,347 @@
+pub struct SskdfDigestAlgorithm {
+    id: SskdfDigestAlgorithmId,
+}
+
+impl SskdfDigestAlgorithm {
+    /// Returns the SSKDF Algorithm Identifier
+    #[must_use]
+    pub fn id(&self) -> SskdfDigestAlgorithmId {
+        self.id
+    }
+
+    #[must_use]
+    fn get_evp_md(&self) -> ConstPointer<EVP_MD> {
+        match_digest_type(match self.id {
+            SskdfDigestAlgorithmId::Sha224 => &AlgorithmID::SHA224,
+            SskdfDigestAlgorithmId::Sha256 => &AlgorithmID::SHA256,
+            SskdfDigestAlgorithmId::Sha384 => &AlgorithmID::SHA384,
+            SskdfDigestAlgorithmId::Sha512 => &AlgorithmID::SHA512,
+        })
+    }
+}
+
+impl PartialEq for SskdfDigestAlgorithm {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl Eq for SskdfDigestAlgorithm {}
+
+impl core::fmt::Debug for SskdfDigestAlgorithm {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        core::fmt::Debug::fmt(&self.id, f)
+    }
+}
+
+/// Single-step (One-step) Key Derivation Function Digest Algorithm Identifier
+#[non_exhaustive]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum SskdfDigestAlgorithmId {
+    /// SSKDF with SHA224
+    Sha224,
+
+    /// SSKDF with SHA256
+    Sha256,
+
+    /// SSKDF with SHA384
+    Sha384,
+
+    /// SSKDF with SHA512
+    Sha512,
+}
+
+/// Single-step (One-step) Key Derivation Function HMAC Algorithm Identifier
+#[non_exhaustive]
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+pub enum SskdfHmacAlgorithmId {
+    /// SSKDF with HMAC-SHA224
+    Sha224,
+
+    /// SSKDF with HMAC-SHA256
+    Sha256,
+
+    /// SSKDF with HMAC-SHA384
+    Sha384,
+
+    /// SSKDF with HMAC-SHA512
+    Sha512,
+}
+
+/// # Single-step Key Derivation Function using HMAC
+///
+/// This algorithm may be referred to as "Single-Step KDF" or "NIST Concatenation KDF" by other
+/// implementors.
+///
+/// ## Implementation Notes
+///
+/// This implementation adheres to the algorithm specified in Section 4 of the
+/// NIST Special Publication 800-56C Revision 2 published on August 2020. The
+/// parameters relevant to the specification are as follows:
+/// * Auxillary Function H is Option 2
+/// * `output.len()`, `secret.len()`, `info.len()` each must be <= 2^30
+/// * `output.len()` and `secret.len()` > 0
+/// * `output.len()`, `secret.len()` are analogous to `L` and `Z` respectively in the
+///   specification.
+/// * `info` refers to `FixedInfo` in the specification.
+/// * `salt.len() == 0` will result in a default salt being used which will be an all-zero byte string
+///   whose length is equal to the length of the specified digest input block length in
+///   bytes.
+///
+/// Specification is available at <https://doi.org/10.6028/NIST.SP.800-56Cr2>
+///
+/// # Errors
+/// `Unspecified` is returned under the following conditions:
+/// * Either `output.len()`, `secret.len()`, or `info.len()` are > 2^30.
+/// * `output.len() == 0 || secret.len() == 0`
```

Comment:
I think we should document the API parameter preconditions & default values above this (outside of "Implementation Details"), and for the "Errors" section just indicate (generically) that `Unspecified` will be returned if any of the parameter precondition checks fail.